pkg/action/executor/limitrange_util.go

```diff
@@ -0,0 +1,317 @@
+package executor
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/golang/glog"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	kclient "k8s.io/client-go/kubernetes"
+)
+
+func CheckLimitrangeViolationOnPod(client *kclient.Clientset, namespace string, pod *corev1.Pod) error {
+	glog.V(4).Infof("Start to check LimitRange violation in the namespace %s based on the pod spec %+v", namespace, pod)
+	if client == nil || pod == nil || namespace == "" {
+		return fmt.Errorf("check failed due to a empty client/pod/namespace")
+	}
+	limitRanges, err := client.CoreV1().LimitRanges(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		glog.Errorf("can't list limitrange due to :%v", err)
+		return err
+	}
+	for _, limitrange := range limitRanges.Items {
+		// Enfource resource requirement against enumerated constraints on the LimitRange
+		// It may modify pod to apply default resource requirements
+		PodMutateLimitFunc(&limitrange, pod)
+		glog.V(4).Infof("The desired podspec becomes %+v after applying the default resource requirements of the limitrange %+v", pod, limitrange)
+		if err := PodValidateLimitFunc(&limitrange, pod); err != nil {
+			return err
+		}
+	}
+	return nil //no limitrange
+}
+
+// PodValidateLimitFunc enforces resource requirements enumerated by the pod against
+// the specified LimitRange.
+func PodValidateLimitFunc(limitRange *corev1.LimitRange, pod *corev1.Pod) error {
+	var errs []error
+
+	for i := range limitRange.Spec.Limits {
+		limit := limitRange.Spec.Limits[i]
+		limitType := limit.Type
+		// enforce container limits
+		if limitType == corev1.LimitTypeContainer {
+			for j := range pod.Spec.Containers {
+				container := &pod.Spec.Containers[j]
+				for k, v := range limit.Min {
+					if err := minConstraint(string(limitType), string(k), v, container.Resources.Requests, container.Resources.Limits); err != nil {
+						errs = append(errs, err)
+					}
+				}
+				for k, v := range limit.Max {
+					if err := maxConstraint(string(limitType), string(k), v, container.Resources.Requests, container.Resources.Limits); err != nil {
+						errs = append(errs, err)
+					}
+				}
+				for k, v := range limit.MaxLimitRequestRatio {
+					if err := limitRequestRatioConstraint(string(limitType), string(k), v, container.Resources.Requests, container.Resources.Limits); err != nil {
+						errs = append(errs, err)
+					}
+				}
+			}
+			for j := range pod.Spec.InitContainers {
```

Comment:
I am not sure if we actually do give resize recommentations for init containers. Is this check needed here?

I can think of a possible environment scenario which can fail our action, even when its not something because of our resize recommendation. A pod was created with a specific limitrange (or not enforceable limit range at all) and the limit range was later created smaller then what the pod is now live with. I am not sure if we should treat that as an actual failure. cc @evat-pm

Reply:
This piece of code is copied from K8s admission plugin. For dealing with `InitContainers`,  it's only doing some comparison and map updates. There is no remote call or deep loop in it. And usually, most pods don't have `InitContainers`.  I kind of leaning towards keeping them if that makes sense to you @irfanurrehman . 
Thanks for your time in reviewing this.

Reply:
ACK!